Implementation of CCPP-ized tropopause_find

jimmielin opened the pull request:
Implements CCPP-compliant tropopause_find.

# Major changes to code
This is my first time doing this, so please feel free to suggest alternative approaches!

1. Standard CCPP-ization procedures, metadata, explicit use of fields instead of physics state, only passing active `:ncol` fields, ... was made.
2. `tropopause_find` accepted "primary" and "backup" methods for finding the tropopause. All the methods known to be used in CAM are now calculated at the same time in the `tropopause_find_run` main driver routine; it calls the underlying logic as appropriate with and populates the appropriate standard name physics fields. Because the *same physical quantities* for tropopause level, height, pressure, temperature, etc. *using different methods* are used throughout the CAM physics and *simultaneously output in the history tape*, the same standard names with a suffix are used to differentiate between these quantities computed by different methods. e.g.,
```
- model_level_number_at_tropopause
- model_level_number_at_tropopause_assuming_climatology
- model_level_number_at_tropopause_assuming_cold_point
- model_level_number_at_tropopause_assuming_hybridstobie_and_climatology
- model_level_number_at_tropopause_assuming_lapse_rate
- model_level_number_at_tropopause_for_chemistry
```
The naming should probably be changed! I'm open to suggestions.

3. `tropopause_read_file`, which provides climatological data used as the default "backup" method to compute the tropopause (`tropp_p_loc` and `tropp_days`), has been moved to a utility module in CAM-SIMA to read, regrid, and provide this climatology to the CCPP-ized physics; it also removes `lchnk` indexing since they're no longer used in CAM-SIMA.
4. tropopause_find used to accept (`optional`) arguments for returning `tropT`, `tropZ`, `tropP` (tropopause temperature, height, and pressure) which are no longer optional per the CAM-SIMA port guide.

A "shim" in current CAM's `tropopause.F90` was made to use the CCPP-ized routines, but preserve all existing behavior and functionality (b4b) as expected in current CAM. In CAM-SIMA all of this are not needed.

# Smaller changes
* `get_curr_calday()` is now provided in the registry as `fractional_calendar_days_on_end_of_current_timestep`

# Outstanding issues/notes

* **Tropopause_find cannot be tested using snapshots within CAM-SIMA:** In regular CAM, `tropopause_find` itself does not update physics state/tendencies; its use in regular CAM is just providing the tropopause level, height, pressure, and temperature to the calling physics package; a final run of `tropopause_output` will then write out to history the computed tropopause properties using the primary method. This means if snapshots cannot verify what `tropopause_find` is doing within CAM-SIMA.

* How to make standard names for the same physical quantities computed using different methods that have to co-exist in one model run

# Answer check
## For existing CAM
* for CAM tested with `QPC6` (which uses CLUBB that uses `tropopause_find`) answers match bit-for-bit for history (`h0i`) and snapshot files before and after `clubb_tend_cam`. I can run more detailed regression tests for the CAM tag.

## For CAM-SIMA
* Note: to test the CAM-SIMA functionality, the history infrastructure should be manually merged in: https://github.com/ESCOMP/CAM-SIMA/pull/274
* Because the tropopause calculated values are only in the history tape and not physics state in CAM, snapshots won't be able to verify answers.


## Tricky way to verify answers
We can validate the answer fields of `TROP_P`, `TROP_T`, and `TROP_Z` can be against the output of CAM by running CAM-SIMA from a CAM snapshot and using history to write out these fields (instantaneously) at the first timestep. Note that `ndens` has to be set to `1` in CAM (for `double`) output, and a `user_set` snapshot should be taken for use before `tropopause_output` in `physpkg.F90` for use in `ncdata`.

nusbaume commented on tropopause_find/tropopause_find.meta:
```diff
@@ -0,0 +1,299 @@
+  dimensions = (horizontal_loop_extent)
+  intent = out
+[ tropZ ]
+  standard_name = tropopause_altitude
```
Just noting that we will need to change the standard names in this file as well, as discussed in previous review comments and in the standard names spreadsheet.

jimmielin replied:
Thanks for the reminder, updated throughout based on CCPP standard names spreadsheet.

nusbaume reviewed (changes requested):
Great work on your first major CCPP-effort @jimmielin!  I have several comments, but they should almost all relate to standard names or changes to code comments.  Of  course please let me know if you have any issues or concerns with anything I requested.  Thanks!

jimmielin commented on cam_diagnostics/tropopause_diagnostics.F90:
```diff
@@ -0,0 +1,215 @@
+    ! Define the output fields.
+    call history_add_field('TROP_P',   'tropopause_air_pressure',    horiz_only,  'avg', 'Pa')
+    call history_add_field('TROP_T',   'tropopause_air_temperature', horiz_only,  'avg', 'K' )
+    call history_add_field('TROP_Z',   'tropopause_altitude',        horiz_only,  'avg', 'm' )
+    call history_add_field('TROP_DZ',  'tropopause_altitude_relative', 'lev',       'avg', 'm')
```
Thank you, names updated according to the CCPP standard names spreadsheet and derivations from these.

jimmielin commented on cam_diagnostics/tropopause_diagnostics.F90:
```diff
@@ -0,0 +1,215 @@
+    call history_add_field('TROP_T',   'tropopause_air_temperature', horiz_only,  'avg', 'K' )
+    call history_add_field('TROP_Z',   'tropopause_altitude',        horiz_only,  'avg', 'm' )
+    call history_add_field('TROP_DZ',  'tropopause_altitude_relative', 'lev',       'avg', 'm')
+    call history_add_field('TROP_PD',  'probability_distribution_of_model_level_number_at_tropopause', 'lev', 'avg', 'probability')
```
Thank you, updated.

jimmielin commented on cam_diagnostics/tropopause_diagnostics.F90:
```diff
@@ -0,0 +1,215 @@
+    call history_add_field('TROPC_Z',  'tropopause_altitude_assuming_climatology',        horiz_only,  'avg', 'm' )
+    call history_add_field('TROPC_DZ', 'tropopause_altitude_relative_assuming_climatology', 'lev', 'avg', 'm')
+    call history_add_field('TROPC_PD', 'probability_distribution_of_model_level_number_at_tropopause_assuming_climatology', 'lev', 'avg', 'probability')
+    call history_add_field('TROPC_FD', 'tropopause_found_assuming_cold_point', horiz_only,  'avg', 'probability')
```
Yes, sharp eyes! Thank you for noticing, fixed.

jimmielin commented on cam_diagnostics/tropopause_diagnostics.F90:
```diff
@@ -0,0 +1,215 @@
+    call history_out_field('hstobie_linoz',  hstobie_linoz)
+    call history_out_field('hstobie_tropop', hstobie_tropop)
+
+    return
```
Thank you, removed here and elsewhere.

jimmielin commented on cam_diagnostics/tropopause_diagnostics.F90:
```diff
@@ -0,0 +1,215 @@
+    call history_add_field('TROPC_FD', 'tropopause_found_assuming_cold_point', horiz_only,  'avg', 'probability')
+
+    ! Hybridstobie output fields
+    call history_add_field('hstobie_trop', 'lower_bound_of_model_level_number_for_stratospheric_chemistry', 'lev', 'inst', 'fraction of model time')
+    call history_add_field('hstobie_linoz', 'lower_bound_of_model_level_number_for_linoz_chemistry', 'lev', 'inst', 'fraction of model time')
+    call history_add_field('hstobie_tropop', 'model_level_number_at_tropopause_for_chemistry', 'lev', 'inst', 'fraction of model time')
```
Thank you, replaced with:
* `hstobie_trop`: `vertical_layer_index_lower_bound_from_hybrid_stobie_linoz_with_climatological_backup_method_for_stratospheric_chemistry`
* `hstobie_linoz`: `vertical_layer_index_lower_bound_from_hybrid_stobie_linoz_with_climatological_backup_method_for_linearized_ozone_chemistry`
* `hstobie_tropop`: `tropopause_vertical_layer_index_from_hybrid_stobie_linoz_with_climatological_backup_method_for_chemistry`

Good question about the units. The original definition (https://github.com/ESCOMP/CAM/blob/29f279540ed497f784b9d73517f5ed828cc6aa10/src/physics/cam/tropopause.F90#L167C1-L170C58) used `fraction of model time` combined with a `I` flag, which is weird to me. I assumed the intent was to `A`verage the fields over time, so you could have a probability distribution averaged over model timesteps of the corresponding hstobie tropopause lower bounds:
```
    call addfld( 'hstobie_trop',   (/ 'lev' /), 'I',  'fraction of model time', 'Lowest level with stratospheric chemsitry')
    call addfld( 'hstobie_linoz',  (/ 'lev' /), 'I',  'fraction of model time', 'Lowest possible Linoz level')
    call addfld( 'hstobie_tropop', (/ 'lev' /), 'I', 'fraction of model time', &
         'Troposphere boundary calculated in chemistry' )
```

I think the code here either has to be updated with `avg`, or the units should be changed.

jimmielin commented on cam_diagnostics/tropopause_diagnostics.meta:
```diff
@@ -0,0 +1,191 @@
+  dimensions = (horizontal_loop_extent, vertical_layer_dimension)
+  intent = in
+[ tropLev ]
+  standard_name = model_level_number_at_tropopause
```
Thanks for sharing! For some reason I missed the `vertical_layer_index` entry in the standard names, so I went digging in the CF conventions (https://cfconventions.org/Data/cf-standard-names/76/build/cf-standard-name-table.html) which has `model_level_number`. `vertical_layer_index` definitely reads better, so I have updated throughout.

jimmielin commented on cam_diagnostics/tropopause_diagnostics.meta:
```diff
@@ -0,0 +1,191 @@
+  dimensions = (horizontal_loop_extent)
+  intent = in
+[ tropLev_twmo ]
+  standard_name = model_level_number_at_tropopause_assuming_lapse_rate
```
Thank you, updated.

jimmielin commented on doc/ChangeLog:
```diff
@@ -1,3 +1,44 @@
+List and Describe any test failures: N/A
+
+Summarize any changes to answers:
+Difference in
```
Thank you, updated! Everything is b4b from my testing. I will be sure to update this, if necessary, after CAM regression tests are ran to ensure b4b to baselines.

jimmielin commented on cam_diagnostics/tropopause_diagnostics.meta:
```diff
@@ -0,0 +1,191 @@
+  dimensions = (horizontal_loop_extent)
+  intent = in
+[ tropLev_hybstob ]
+  standard_name = model_level_number_at_tropopause_assuming_hybridstobie_and_climatology
```
Thank you, updated.

jimmielin commented on tropopause_find/tropopause_find.F90:
```diff
@@ -0,0 +1,1493 @@
+
+  save
+
+  ! These parameters define and enumeration to be used to define the primary
+  ! and backup algorithms to be used with the tropopause_find() method. The
+  ! backup algorithm is meant to provide a solution when the primary algorithm
+  ! fail. The algorithms that can't fail are: TROP_ALG_ANALYTIC, TROP_ALG_CLIMATE
+  ! and TROP_ALG_STOBIE.
```
Thank you, updated.

jimmielin commented on tropopause_find/tropopause_find.F90:
```diff
@@ -0,0 +1,1493 @@
+! when the original algorithm fails. The tropopause temperature and
+! pressure are determined and can be output to the history file.
+!
+! These routines are based upon code in the WACCM chemistry module
+! including mo_tropoause.F90 and llnl_set_chem_trop.F90. The code
+! for the Reichler et al. [2003] algorithm is from:
+!
+!   http://www.gfdl.noaa.gov/~tjr/TROPO/tropocode.htm
+!
```
Thanks, I have updated to point to the DOI with the scientific citation (Reichler et al. 2003): Reichler, T., M. Dameris, and R. Sausen (2003), Determining the tropopause height from gridded data, Geophys. Res. Lett., 30, 2042, doi:[10.1029/2003GL018240](https://doi.org/10.1029/2003GL018240), 20.

jimmielin commented on tropopause_find/tropopause_find.F90:
```diff
@@ -0,0 +1,1493 @@
+contains
+!================================================================================================
+
+  ! This routine is called during intialization and must be called before the
+  ! other methods in this module can be used. Its main tasks are to read in the
+  ! climatology from a file and to define the output fields. Much of this code
+  ! is taken from mo_tropopause.
```
Thanks, removed.

jimmielin commented on tropopause_find/tropopause_find.F90:
```diff
@@ -0,0 +1,1493 @@
+
+  real(kind_phys), parameter :: ALPHA  = 0.03_kind_phys
+
+  ! FIXME hplin 8/15/24: fillvalue from cam_history_support. To check how it is used
+  ! and if we can remove it with a generic (NOTFOUND?) value
+  real(kind_phys), parameter :: fillvalue = 1.e36_kind_phys
```
1. Thanks for the suggestion, I assume if I define it in `src/data/registry.xml` I'd have to define the value somewhere then mark it as initialized. Where would be the best place to assign this? Or can I assign a default value from the registry xml file?

I was wondering if it would be possible to add `.meta` to `cam_history_support.F90` (https://github.com/ESCOMP/CAM-SIMA/blob/0784bcfc66bfdca27af9627b27253b580ffa6694/src/history/cam_history_support.F90) to pass `[fillvalue]`, similar to `tropopause_read_file` passes the climatology?

2. Would `fill_value_for_diagnostic_output` be an appropriate standard name for this?

jimmielin commented on tropopause_find/tropopause_find.F90:
```diff
@@ -0,0 +1,1493 @@
+    integer,         intent(in)         :: ncol          ! Number of atmospheric columns
+    integer,         intent(in)         :: pver          ! Number of vertical levels
+    real(kind_phys), intent(in)         :: lat(:)        ! Latitudes (radians)
+    real(kind_phys), intent(in)         :: pint(:,:)     ! Interface pressures (Pa), pverp
```
Thank you, removed.

jimmielin commented on tropopause_find/tropopause_find.F90:
```diff
@@ -0,0 +1,1493 @@
+    real(kind_phys), intent(in)         :: pint(:,:)     ! Interface pressures (Pa), pverp
+    real(kind_phys), intent(in)         :: pmid(:,:)     ! Midpoint pressures (Pa)
+    real(kind_phys), intent(in)         :: t(:,:)        ! Temperature (K)
+    real(kind_phys), intent(in)         :: zi(:,:)       ! Geopotential height above surface at interfaces (m), pverp
+    real(kind_phys), intent(in)         :: zm(:,:)       ! Geopotential height above surface at midpoints (m), pver
```
Thank you, removed throughout.

jimmielin commented on tropopause_find/tropopause_find.F90:
```diff
@@ -0,0 +1,1493 @@
+    real(kind_phys), intent(in)         :: zm(:,:)       ! Geopotential height above surface at midpoints (m), pver
+    real(kind_phys), intent(in)         :: phis(:)       ! Surface geopotential (m2 s-2)
+
+    real(kind_phys), intent(in)         :: calday        ! Day of year including fraction from get_curr_calday
```
Thank you, updated throughout.

jimmielin commented on tropopause_find/tropopause_find.F90:
```diff
@@ -0,0 +1,1493 @@
+    real(kind_phys), intent(in)         :: calday        ! Day of year including fraction from get_curr_calday
+
+    ! Climatological tropopause pressures (Pa), (pcols,ntimes=12).
+    ! Remark: Not chunkized, subsetted to chunk for backwards compatibility with CAM
```
Thank you, agreed, the corresponding call in CAM has a similar remark so I will just remove it here:

```
         tropp_p_loc    = tropp_p_loc(:,pstate%lchnk,:), &  ! Subset into chunk as the underlying routines are no longer chunkized.
```

jimmielin commented on tropopause_find/tropopause_find.F90:
```diff
@@ -0,0 +1,1493 @@
+    end do
+  end subroutine tropopause_analytic
+
+  ! Read the tropopause pressure in from a file containging a climatology. The
+  ! data is interpolated to the current dat of year and latitude.
+  !
+  ! NOTE: The data is read in during tropopause_init and stored in the module
+  ! variable trop
```
Thank you, updated.

jimmielin commented on tropopause_find/tropopause_find.F90:
```diff
@@ -0,0 +1,1493 @@
+  subroutine tropopause_hybridstobie(ncol, pver, pmid, t, zm, &
+                                     tropLev, tropP, tropT, tropZ, &
+                                     hstobie_trop, hstobie_linoz, hstobie_tropop)
+    !use cam_history,  only : outfld
```
Thank you, removed.

jimmielin commented on tropopause_find/tropopause_find.F90:
```diff
@@ -0,0 +1,1493 @@
+      hstobie_tropop(:,:) = trop_trop_output(:,:)
+    endif
+
+    !call outfld( 'hstobie_trop',   trop_output(:ncol,:),       ncol, pstate%lchnk )
+    !call outfld( 'hstobie_linoz',  trop_linoz_output(:ncol,:), ncol, pstate%lchnk )
+    !call outfld( 'hstobie_tropop', trop_trop_output(:ncol,:),  ncol, pstate%lchnk )
```
Thank you, removed.

jimmielin commented on tropopause_find/tropopause_find.F90:
```diff
@@ -0,0 +1,1493 @@
+  !
+  !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+  !
+  ! determination of tropopause height from gridded temperature data
+  !
+  ! reference: Reichler, T., M. Dameris, and R. Sausen (2003)
```
Thank you, updated.

jimmielin commented on test/test_sdfs/suite_tropopause_find.xml:
```diff
@@ -0,0 +1,12 @@
+
+<suite name="tropopause_find" version="1.0">
+  <group name="physics_before_coupler">
+    <!-- Find tropopause -->
+    <!-- Tropopause find has to make history output at the end of convective adjustment -->
+    <!-- so supersaturated cells are not written to the tape. -->
+    <scheme>tropopause_find</scheme>
+    <scheme>tropopause_diagnostics</scheme>
```
Thank you, added.

jimmielin commented on cam_diagnostics/tropopause_diagnostics.meta:
```diff
@@ -0,0 +1,191 @@
+  intent = in
+[ tropLev ]
+  standard_name = model_level_number_at_tropopause
+  units = 1
```
Thank you, updated throughout.

jimmielin commented on tropopause_find/tropopause_find.F90:
```diff
@@ -0,0 +1,1493 @@
+    return
+  end subroutine tropopause_twmo
+  
+  ! This routine implements the WMO definition of the tropopause (WMO, 1957; Seidel and Randel, 2006).
```
Thank you, citation added.

jimmielin commented on tropopause_find/tropopause_find.F90:
```diff
@@ -0,0 +1,1493 @@
+  ! backup algorithm is meant to provide a solution when the primary algorithm
+  ! fail. The algorithms that can't fail are: TROP_ALG_ANALYTIC, TROP_ALG_CLIMATE
+  ! and TROP_ALG_STOBIE.
+  integer, parameter    :: TROP_ALG_NONE      = 1    ! Don't evaluate
+  integer, parameter    :: TROP_ALG_ANALYTIC  = 2    ! Analytic Expression
+  integer, parameter    :: TROP_ALG_CLIMATE   = 3    ! Climatology
+  integer, parameter    :: TROP_ALG_STOBIE    = 4    ! Stobie Algorithm
+  integer, parameter    :: TROP_ALG_TWMO      = 5    ! WMO Definition, Reichler et al. [2003]
+  integer, parameter    :: TROP_ALG_WMO       = 6    ! WMO Definition
+  integer, parameter    :: TROP_ALG_HYBSTOB   = 7    ! Hybrid Stobie Algorithm
+  integer, parameter    :: TROP_ALG_CPP       = 8    ! Cold Point Parabolic
+  integer, parameter    :: TROP_ALG_CHEMTROP  = 9    ! Chemical tropopause
```
Thank you for the suggestion, I've posted https://github.com/ESCOMP/atmospheric_physics/issues/116

jimmielin commented:
Updated history field list with new CCPP field names:
```
  ***************** HISTORY FIELD LIST ******************
          TROP_P           Pa  avg  tropopause_air_pressure
          TROP_T            K  avg  tropopause_air_temperature
          TROP_Z            m  avg  tropopause_geopotential_height_wrt_surface
         TROP_DZ            m  avg  geopotential_height_difference_between_atmosphere_layer_and_tropopause
         TROP_PD  probability  avg  probability_distribution_of_tropopause_vertical_layer_index
         TROP_FD  probability  avg  tropopause_found
         TROPP_P           Pa  avg  tropopause_air_pressure_from_lapse_rate_method
         TROPP_T            K  avg  tropopause_air_temperature_from_lapse_rate_method
         TROPP_Z            m  avg  tropopause_geopotential_height_wrt_surface_from_lapse_rate_method
        TROPP_DZ            m  avg  geopotential_height_difference_between_atmosphere_layer_and_tropopause_from_lapse_rate_method
        TROPP_PD  probability  avg  probability_distribution_of_tropopause_vertical_layer_index_from_lapse_rate_method
        TROPP_FD  probability  avg  tropopause_found_from_lapse_rate_method
         TROPF_P           Pa  avg  tropopause_air_pressure_from_cold_point_method
         TROPF_T            K  avg  tropopause_air_temperature_from_cold_point_method
         TROPF_Z            m  avg  tropopause_geopotential_height_wrt_surface_from_cold_point_method
        TROPF_DZ            m  avg  geopotential_height_difference_between_atmosphere_layer_and_tropopause_from_cold_point_method
        TROPF_PD  probability  avg  probability_distribution_of_tropopause_vertical_layer_index_from_cold_point_method
        TROPF_FD  probability  avg  tropopause_found_from_cold_point_method
         TROPC_P           Pa  avg  tropopause_air_pressure_from_climatological_method
         TROPC_T            K  avg  tropopause_air_temperature_from_climatological_method
         TROPC_Z            m  avg  tropopause_geopotential_height_wrt_surface_from_climatological_method
        TROPC_DZ            m  avg  geopotential_height_difference_between_atmosphere_layer_and_tropopause_from_climatological_method
        TROPC_PD  probability  avg  probability_distribution_of_tropopause_vertical_layer_index_from_climatological_method
        TROPC_FD  probability  avg  tropopause_found_from_climatological_method
    hstobie_trop fraction of   ins  vertical_layer_index_lower_bound_from_hybrid_stobie_linoz_with_climatological_backup_method_for_stratospheric_chemistry
   hstobie_linoz fraction of   ins  vertical_layer_index_lower_bound_from_hybrid_stobie_linoz_with_climatological_backup_method_for_linearized_ozone_chemistry
  hstobie_tropop fraction of   ins  tropopause_vertical_layer_index_from_hybrid_stobie_linoz_with_climatological_backup_method_for_chemistry
  *************** END HISTORY FIELD LIST ****************
```

jimmielin commented on tropopause_find/tropopause_find.meta:
```diff
@@ -0,0 +1,299 @@
+  dimensions = ()
+  intent = in
+[ tropp_p_loc ]
+  standard_name = tropopause_air_pressure_from_climatology
```
This is the input climatological dataset from `tropopause_read_file`. Does this standard name look fine? I am a little concerned it looks too similar to `tropopause_air_pressure_from_climatological_method`, so I was thinking if it would be good to change to `tropopause_air_pressure_from_climatology_dataset`.

cacraigucar replied:
I like using `tropopause_air_pressure_from_climatology_dataset`

jimmielin commented on cam_diagnostics/tropopause_diagnostics.meta:
```diff
@@ -0,0 +1,191 @@
+  dimensions = (horizontal_loop_extent)
+  intent = in
+[ tropZ ]
+  standard_name = tropopause_altitude
```
Thanks, updated throughout.

jimmielin commented:
Thanks @nusbaume for your review comments. I have updated standard names and updated comments throughout to address your comments.

Tests on Derecho show changes here are b4b with the original PR.

There is one exception for `fillvalue` as I'm not quite sure where we would set its value after defining it in the registry for it to be available in both the CCPP physics and `cam_history_support`. I noted an alternative approach in the PR comment thread. Could you and @peverwhee perhaps provide some additional guidance? Thank you!

nusbaume reviewed (approved):
Thanks for implementing all of my change requests!  There was one last comment change request I missed the first time, but otherwise everything looks good (so no need to request a re-review).  Thanks!

jimmielin commented:
Thanks @nusbaume for the review! I've committed your suggested change.

Can I push the big green button to merge in the pull request into `development` now or are there any steps to complete? https://github.com/ESCOMP/atmospheric_physics/wiki/Tagging-Instructions mentions a tag should be made as well but I don't see any development tags, tags seem to be made based on `main`. Thanks!

cacraigucar commented on doc/ChangeLog:
```diff
@@ -1,3 +1,43 @@
+    - tropopause_find accepted "primary" and "backup" methods for finding the tropopause. All the methods known to be used in CAM are now calculated at the same time in the tropopause_find_run main driver routine; it calls the underlying logic as appropriate with and populates the appropriate standard name physics fields. Because the same physical quantities for tropopause level, height, pressure ...
+    - Standard CCPP-ization procedures, metadata, explicit use of fields instead of physics state, only passing active :ncol fields, ...
+
+    A "shim" that will completely replicate existing behavior, bit-for-bit in current CAM, will be included in current CAM as tropopause.F90.
```
We typically use "interface code" in CAM to signify what I believe you are conveying with the word "shim".  If that is equivalent, then there was at least one other place where I saw the word "shim"

jimmielin replied:
Thank you, updated with "interface code". I believe the other place was in the ESCOMP/CAM pull request, which I've now updated as well.

cacraigucar reviewed (changes requested):
Impressive conversion.  Great job!  

We'll need to make sure this code is tested in ESCOMP/CAM regression tests before it can be committed into this repo.

jimmielin commented on tropopause_find/tropopause_find.F90:
```diff
@@ -0,0 +1,1464 @@
+! when the original algorithm fails. The tropopause temperature and
+! pressure are determined and can be output to the history file.
+!
+! Author: Charles Bardeen
+! Created: April, 2009
+!
+! CCPP-ized: Haipeng Lin, August 2024
```
Thank you for your comments 😄

I am happy to keep/remove. I looked through files and could not find a standard header for CCPP-ized module files. Perhaps we could have one?

nusbaume replied:
I don't think any of us have particularly strong opinions here, but in general we lean towards not including authors and dates because it can make the comment headers excessively long, and that information can usually be extracted from the git log and/or `ChangeLog` already.

jimmielin commented on tropopause_find/tropopause_find.F90:
```diff
@@ -0,0 +1,1464 @@
+
+  end subroutine tropopause_cpp
+
+  ! Searches all the columns in the chunk and attempts to identify the "chemical"
```
Removed, thank you!

jimmielin commented on tropopause_find/tropopause_find.F90:
```diff
@@ -0,0 +1,1464 @@
+    real(kind_phys), intent(in)         :: calday        ! Day of year including fraction of day
+
+    ! Climatological tropopause pressures (Pa), (pcols,ntimes=12).
+    ! Remark: Not chunkized, subsetted to chunk for backwards compatibility with CAM
```
Removed, thank you!

jimmielin commented on tropopause_find/tropopause_find.F90:
```diff
@@ -0,0 +1,1464 @@
+    real(kind_phys), intent(in)         :: calday        ! Day of year including fraction of day
+
+    ! Climatological tropopause pressures (Pa), (pcols,ntimes=12).
+    ! Remark: Not chunkized, subsetted to chunk for backwards compatibility with CAM
```
Removed, thank you!

jimmielin commented on tropopause_find/tropopause_find.F90:
```diff
@@ -0,0 +1,1464 @@
+    real(kind_phys), intent(in)         :: calday        ! Day of year including fraction of day
+
+    ! Climatological tropopause pressures (Pa), (pcols,ntimes=12).
+    ! Remark: Not chunkized, subsetted to chunk for backwards compatibility with CAM
```
Removed, thank you!

jimmielin commented on tropopause_find/tropopause_find.F90:
```diff
@@ -0,0 +1,1464 @@
+
+  end subroutine tropopause_find_run
+
+  ! Searches all the columns in the chunk and attempts to identify the tropopause.
```
Removed, thank you!

jimmielin commented on cam_diagnostics/tropopause_diagnostics.F90:
```diff
@@ -0,0 +1,220 @@
+  public :: tropopause_diagnostics_init
+  public :: tropopause_diagnostics_run
+
+  ! Parameters consistent with tropopause_find - can they be imported?
```
I removed the comment. I could export a CCPP standard quantity for `NOTFOUND = -1` (`flag_for_tropopause_not_found`? or a more generic term for invalid indices generally) if other schemes want to check for it.

jimmielin commented on cam_diagnostics/tropopause_diagnostics.F90:
```diff
@@ -0,0 +1,220 @@
+
+    return
+  end subroutine tropopause_diagnostics_run
+end module tropopause_diagnostics
```
Fixed, thank you!

jimmielin commented on cam_diagnostics/tropopause_diagnostics.F90:
```diff
@@ -0,0 +1,220 @@
+    call history_out_field('hstobie_linoz',  hstobie_linoz)
+    call history_out_field('hstobie_tropop', hstobie_tropop)
+
+    return
```
Oops, missed one. Removed, thanks!

jimmielin commented on cam_diagnostics/tropopause_diagnostics.F90:
```diff
@@ -0,0 +1,220 @@
+    call history_add_field('TROPF_PD', 'probability_distribution_of_tropopause_vertical_layer_index_from_cold_point_method', 'lev', 'avg', 'probability')
+    call history_add_field('TROPF_FD', 'tropopause_found_from_cold_point_method', horiz_only,  'avg', 'probability')
+
+    ! Climatology only - will never fail
```
I agree it's confusing, I've changed it to "will always find a tropopause", which is what "not fail" means in the original `tropopause.F90`. I've also defined this more clearly in the `tropopause_find.F90` comments for clarity:

```
  ! backup algorithm is meant to provide a solution when the primary algorithm
  ! fails. The algorithms that can't fail (i.e., always find a tropopause) are:
  ! TROP_ALG_ANALYTIC, TROP_ALG_CLIMATE, and TROP_ALG_STOBIE.
```

jimmielin commented on cam_diagnostics/tropopause_diagnostics.F90:
```diff
@@ -0,0 +1,220 @@
+    call history_add_field('TROPC_T',  'tropopause_air_temperature_from_climatological_method', horiz_only,  'avg', 'K' )
+    call history_add_field('TROPC_Z',  'tropopause_geopotential_height_wrt_surface_from_climatological_method', horiz_only, 'avg', 'm' )
+    call history_add_field('TROPC_DZ', 'geopotential_height_difference_between_atmosphere_layer_and_tropopause_from_climatological_method', 'lev', 'avg', 'm')
+    call history_add_field('TROPC_PD', 'probability_distribution_of_tropopause_vertical_layer_index_from_climatological_method', 'lev', 'avg', 'probability')
+    call history_add_field('TROPC_FD', 'tropopause_found_from_climatological_method', horiz_only,  'avg', 'probability')
```
Thank you for the great question! I scanned throughout the CAM code and these are the methods that are used in the current version of the code. Others in CAM were put behind a logical to output all methods, but it was hardcoded in the `tropopause.F90` code as `.false.`. Because of the (a little unfortunate) duplication of code / standard names when adding a new method I supported only the ones that are currently used, and we have a cleanup issue (#116) to remove the unused methods eventually.

cacraigucar commented on tropopause_find/tropopause_find.F90:
```diff
@@ -0,0 +1,1464 @@
+
+    real(kind_phys), intent(in)         :: calday        ! Day of year including fraction of day
+
+    ! Climatological tropopause pressures (Pa), (pcols,ntimes=12).
```
As I was resolving comments, I saw this is comment which refers to pcols.  It (and any place else which refers to `pcols `in this code) should be replaced with `ncol`.

jimmielin replied:
Thank you, replaced with ncol!

cacraigucar reviewed (changes requested):
Minor mods requested

jimmielin commented:
Thanks @cacraigucar, comment updated (also in CAM-SIMA) and standard name for `tropp_p_loc` changed to `tropopause_air_pressure_from_climatology_dataset`.

jimmielin commented:
Hi @cacraigucar I ran the regression tests with the CCPP-ized tropopause_find. Tests on Izumi NAG match pre-existing failures; GNU all passed as expected. All tests on Derecho match pre-existing failures except for the SMS `FCHIST_GC` test:

```
➜  aux_cam_intel_20240925232609 ./cs.status.aux_cam_intel_20240925232609 | testgrep
aux_cam_intel_20240925232609: 63 tests
  ERP_Ln9.f09_f09_mg17.FCSD_HCO.derecho_intel.cam-outfrq9s (Overall: FAIL) details:
    FAIL ERP_Ln9.f09_f09_mg17.FCSD_HCO.derecho_intel.cam-outfrq9s COMPARE_base_rest
  SMS_D_Ln9_P1280x1.ne0CONUSne30x8_ne0CONUSne30x8_mt12.FCHIST.derecho_intel.cam-outfrq9s (Overall: PEND) details:
    PEND SMS_D_Ln9_P1280x1.ne0CONUSne30x8_ne0CONUSne30x8_mt12.FCHIST.derecho_intel.cam-outfrq9s SHAREDLIB_BUILD
  SMS_Ld1.f09_f09_mg17.FCHIST_GC.derecho_intel.cam-outfrq1d (Overall: DIFF) details:
    FAIL SMS_Ld1.f09_f09_mg17.FCHIST_GC.derecho_intel.cam-outfrq1d BASELINE /glade/campaign/cesm/community/amwg/cam_baselines/cam6_4_035_intel: DIFF
```

Upon further investigation of the `FCHIST_GC` test output, the results of the `FCHIST_GC` match the baselines of `cam6_4_034_intel` (and `cam6_4_036_intel`) but not `cam6_4_035_intel`; the `cam6_4_035` changelog marks `FCHIST_GC` as a pre-existing failure but I'm not aware of the SMS test for FCHIST_GC failing before. Maybe there was a machine hiccup when running the cam6_4_035 tests that resulted in a DIFF.

When I ran this only `cam6_4_035` was tagged and not `cam6_4_036`. Given that we will run the regression tests again when merging the CAM PR, and results on cam6_4_035 look as expected, I would suggest we could merge into `atmospheric_physics` first then run the CAM tests again when ready for a tag?

Thanks!
